### DISCUSSION TITLE
What functionality are you missing in the framework?

### DISCUSSION THREAD

**Discussion by reindert-vetter** [confetti-framework/confetti#97] [Ideas]
What functionality are you missing in the framework? I am now thinking of building [Rate Limiting](https://laravel.com/docs/8.x/routing#rate-limiting). But what do you need? _Functions are fully implemented in one go as much as possible._

**Comment by Wulfheart** [comment]
I really like the Livewire stuff in Laravel but this is more a package for userland.
Components in templates would be nice though.

Furthermore:

- Auth scaffolding (maybe something like Fortify?)

**Comment by NWBY** [comment]
Rate limiting seems like a good idea. A few things that come to mind are: 
- Custom helpers
- Webpack support (I believe this could easily be achieved by using Laravel Mix)
- DB and ORM setup

Apologies if any of these have already been implemented.

**Reply by Wulfheart** [reply]
For ORM we should take a look at [Prisma](https://github.com/prisma/prisma-client-go).

**Reply by reindert-vetter** [reply]
**ORM:** I've thought about that for a long time. But I think I am waiting for generics in Go. I can make all subjects in such a way that gives the same feeling as Laravel, only an ORM needs generics in my opinion.
Until then you can use https://gorm.io/ (or Prisma), for example.

**Reply by reindert-vetter** [reply]
> * Custom helpers

Which helpers do you miss? Go already has quite a few 'helper' packages.

**Reply by rafaelbreno** [reply]
One of the best helpers that I often use is the [tap()](https://laravel.com/docs/8.x/helpers#method-tap) functions, I've seen some approaches in Go, [here's](https://play.golang.org/p/WOnE31ldpb) one of them.

**Reply by reindert-vetter** [reply]
That looks good. Thanks! I have created a new discussion for custom helpers:
https://github.com/confetti-framework/confetti/discussions/102

**Reply by reindert-vetter** [reply]
> * Webpack support

New discussion:
https://github.com/confetti-framework/confetti/discussions/104

**Reply by Wulfheart** [reply]
Any ETA for generics in Go?

**Reply by reindert-vetter** [reply]
@Wulfheart ~End of this year.

**Reply by reindert-vetter** [reply]
New ORM discussion: https://github.com/confetti-framework/confetti/discussions/110

**Comment by rafaelbreno** [comment]
- The validation section could be more in depth
    - E.g. the "unique" key, if you open a issue I could help adding more validations, implementing TDD in those new features
- CMD commands to generate pre-defined files 
- Queues/Jobs

**Reply by reindert-vetter** [reply]
Great that you want to help! Besides the documentation you can obviously see how other rules are implemented.
You can use issue https://github.com/confetti-framework/validation/issues/1

**Reply by almas-x** [reply]
> Great that you want to help! Besides the documentation you can obviously see how other rules are implemented.
> You can use issue [confetti-framework/validation#1](https://github.com/confetti-framework/validation/issues/1)

Speaking of validation, Is there any plan to add `Validate` method for `request` ?
example
```go
failures := request.Validate(
    val.Verify("name", rule.Required{}, rule.String{}),
    val.Verify("email", rule.Required{}, rule.Email{}),
)
```

**Reply by reindert-vetter** [reply]
That is nice indeed, but it is not possible to validate the parameters in that way. But maybe that's not so bad.

I also wanted to make the interface of `request` as small as possible (it will be easier to mock). But I can see that this is very useful. What do others think about this?

**Comment by bosix** [comment]
Scheduling and async job execution are really great in laravel (https://laravel.com/docs/8.x/scheduling). Would love to see it in confetti.

I suggest a queue-worker system for job execution that utilizes goroutines for parallel execution and a worker pool to prevent too many jobs from being executed at the same time.

**Comment by almas-x** [comment]
How about anonymous validation function, easy to define custom validation rules
example

```go
type VerifyFunc func(value support.Value) error

func (f VerifyFunc) Verify(value support.Value) error {
	return f(value)
}

failures := val.Validate(app, content,
		val.Verify("some.field", val.VerifyFunc(func(value support.Value) error {
                        // do awesome validation
			return nil
		})))
```

**Reply by reindert-vetter** [reply]
That is indeed a very good idea and easy to implement. Way not:

```
failures := val.Validate(app, content,
val.Verify("some.field", func(value support.Value) error {
         // do awesome validation
	return nil
}))
```
Then that interface must be made smaller, but than it is the same as Laravel.

https://laravel.com/docs/8.x/validation#using-closures

**Reply by almas-x** [reply]
Because that `val.Verify` received a `inter.Rule` interface, not the anonymous function. However `val.VerifyFunc` can convert 
anonymous function to `inter.Rule` interface. In this way you don't need change any code in `val.Verify` method.

**Comment by nurmuhammet-ali** [comment]
Could be better if documentation had "Deployment" block
